### PR TITLE
RequireJS

### PR DESCRIPTION
**Pull request by esoergel**
Work-in-progress of requirejs integration.  I'm looking for feedback, hoping discussion can happen early so if there's anything we don't like, we can change it.

### CODE REVIEW

**Comment by esoergel** [comment]
@dimagi/js-team take a look at this PR and let me know if you have any feedback.  @millerdev if you're up for it, I'd be interested in seeing what a similar PR looks like with CommonJS/Browserify.

File: corehq/apps/hqwebapp/static/hqwebapp/js/hq_paths.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        "underscore-mixins": "/static/style/js/underscore-mixins",
+
+    },
+    shim: {
+        "jquery.form": ["jquery"],
+        "bootstrap": ["jquery"],
+        "jquery.cookie": ["jquery"],
+        "jquery.hq": ["jquery"],
+    },
+});
```
Comment (esoergel) [inline corehq/apps/hqwebapp/static/hqwebapp/js/hq_paths.js]: As of right now, this is a duplicate of the file in `base.html`, I stuck it in there for now for convenience, but long term, it should ideally be something generated once and served as a static file.  Maybe this process could be hooked in to `collectstatic`?

File: corehq/apps/locations/static/locations/ko/location_types.js [hunk 2]
```diff
@@ -161,3 +163,8 @@ ko.bindingHandlers.bind_element = {
         }
     }
 };
+
+    return {
+        LocationSettingsViewModel: LocationSettingsViewModel,
+    };
+});
```
Comment (esoergel) [inline corehq/apps/locations/static/locations/ko/location_types.js]: This is all it takes per-file to convert existing files to use `require.js`.

File: corehq/apps/style/static/style/js/hq.helpers.js [hunk 3]
```diff
@@ -91,43 +91,3 @@ $.showMessage = function (message, level) {
     $notice.append(message);
     $(".hq-page-header-container").prepend($notice);
 };
```
Comment (esoergel) [inline corehq/apps/style/static/style/js/hq.helpers.js]: I moved the jquery additions to `hq_extensions.jquery.js` so they're all in one place.  I was having some difficulty with the other ones, so I just kept 'em here for now.

File: corehq/apps/style/templates/style/bootstrap2/base.html [hunk 4]
```diff
@@ -392,6 +392,39 @@ <h1 class="page-header">
...
+                    "jquery": "{% new_static 'jquery/dist/jquery.min' %}",
+                    "bootstrap": "{% new_static 'style/lib/bootstrap-3.2.0/dist/js/bootstrap.min' %}",
+                    "jquery.form": "{% new_static 'jquery-form/jquery.form' %}",
+                    "jquery.cookie": "{% new_static 'jquery.cookie/jquery.cookie' %}",
+                    "jquery.hq": "{% new_static 'style/js/hq_extensions.jquery' %}",
+
+                    "jquery17": "{% new_static 'jquery-1.7.1-legacy/jquery' %}",
+                    "knockout": "{% new_static 'knockout/dist/knockout' %}",
+                    "underscore": "{% new_static 'underscore/underscore' %}",
+                    "underscore-mixins": "{% new_static 'style/js/underscore-mixins' %}",
```
Comment (esoergel) [inline corehq/apps/style/templates/style/bootstrap2/base.html]: The tricky thing about our system is that we need `{% static %}` tags around our resources.  Right now, you can only reference another file from within the html page (where `static` is available), or by defining it in the config.  You might be able to define a second config in the necessary html page, and then reference it from within js files.

Alternatively, if we can get rid of the versioning on js files (so you can reference a file as `{{ STATIC_ROOT }}/jquery/dist/jquery.min.js`), we could just specify a `path` variable in this config and refence everything relative to that:

```javascript
require(['jquery/dist/jquery.min'], function($) {});
```

File: corehq/apps/hqwebapp/static/hqwebapp/js/hquery.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+// bundles jquery plugins with jquery in one dependency
```
Comment (orangejenny) [inline corehq/apps/hqwebapp/static/hqwebapp/js/hquery.js]: picky, but I'd name this something else. jquery_extended? jquery_hq?

File: corehq/apps/locations/static/locations/ko/location_types.js [hunk 6]
```diff
@@ -161,3 +163,8 @@ ko.bindingHandlers.bind_element = {
         }
     }
 };
+
+    return {
```
Comment (orangejenny) [inline corehq/apps/locations/static/locations/ko/location_types.js]: For this specific file, it'd be less verbose to make LocationSettingsViewModel an anonymous function and return that function rather than a hash. Then in location_types.html, instead of `function($, ko, location_types) {` you'd use `function($, ko, LocationSettingsViewModel) {` and instead of `new location_types.LocationSettingsViewModel(loc_types, commtrack_enabled);` you could leave it as `new LocationSettingsViewModel(loc_types, commtrack_enabled);` (I'd also likely rename this file to location_settings_view_model.js.) Thoughts?

**Comment by czue** [comment]
cool!

**Comment by esoergel** [comment]
Noting that I should check for new usages of `hq.helpers.js` before this gets merged, there's [one in PR](https://github.com/dimagi/commcare-hq/pull/9614/files#r47813756) now that'd be broken by this.

File: corehq/apps/locations/static/locations/ko/location_types.js (hunk 6 above)
Comment (esoergel) [inline corehq/apps/locations/static/locations/ko/location_types.js]: That's a good point, and it sounds like it'd work.  I'm hesitant though, because I've been thinking of this in terms of modules, where the object returned is a namespace.

I could probably be talked out of it, but I'm inclined to make the define function always return a module, just for consistency.

File: corehq/apps/hqwebapp/static/hqwebapp/js/hquery.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+// otherwise we'd need to specify which ones are required each time
+// which is nice and explicit, but makes it harder to adapt existing code
+define([
+    "jquery",
+    "jquery.form",
+    "bootstrap",
```
Comment (proteusvacuum) [inline corehq/apps/hqwebapp/static/hqwebapp/js/hquery.js]: does this count as a jquery plugin? bit confusing to see this here otherwise.

File: corehq/apps/style/templates/style/bootstrap2/base.html (hunk 4 above)
Comment (benrudolph) [inline corehq/apps/style/templates/style/bootstrap2/base.html]: i like the idea of a path variable. though seems like it'd be hard in our case cause we have so many paths. every app + bower

File: corehq/apps/style/static/style/js/hq_extensions.jquery.js [hunk 8]
```diff
@@ -44,9 +44,43 @@
...
+        $(this).addSpinnerToButton();
+    };
+
+    $.fn.enableButton = function () {
+        $(this).removeSpinnerFromButton();
+        $(this).removeClass('disabled')
+            .removeAttr('disabled');
+    };
+
+    $.fn.koApplyBindings = function (context) {
```
Comment (proteusvacuum) [inline corehq/apps/style/static/style/js/hq_extensions.jquery.js]: Haven't been part of the discussions for this, so I might be overthinking this -- does using require not allow us to be much more selective than having to import whole util files when we want to just use a single function that is within it? I'm thinking of this one (`koApplyBindings`) in particular -- it seems that this will be used in most places we use knockout, where I might not care about disabling buttons with spinners. 
Basically, I'm wondering if it is worth breaking these files out more so we are more explicit in what exactly we are using?

**Comment by proteusvacuum** [comment]
Very excited about this!

File: corehq/apps/style/templates/style/bootstrap2/base.html (hunk 4 above)
Comment (esoergel) [inline corehq/apps/style/templates/style/bootstrap2/base.html]: We don't necessarily need to do every path - it could be just `<base_url>/static/`, as everything can be defined relative to that.  The main problem is that `{% static %}` does more than add a prefix (it appends a version ID), so that wouldn't work.  Changing that is definitely outside the scope of this project though.

File: corehq/apps/hqwebapp/static/hqwebapp/js/hquery.js (hunk 7 above)
Comment (esoergel) [inline corehq/apps/hqwebapp/static/hqwebapp/js/hquery.js]: Yup, you access bootstrap's JS functions via `$("#id").fn`, so it needs to hook in to jquery.

File: corehq/apps/style/static/style/js/hq_extensions.jquery.js (hunk 8 above)
Comment (esoergel) [inline corehq/apps/style/static/style/js/hq_extensions.jquery.js]: You definitely can do that if you break these files apart.  Here's what that might look like:
```javascript
// file 1
define(['jquery', 'jquery.form', 'bootstrap'], function($, ignore1, ignore2) {});
// file 2
define(['jquery', 'bootstrap', 'jquery.cookie', 'jquery.hq'], function($, ignore1, ignore2, ignore3) {});
```
The chief advantage here is that you can be as granular as you like with resources.

The way I've implemented it is as a monolith, so in both cases it'd just be 
```javascript
define(['hquery'], function($) {});
```
The advantages of this are:
* simpler to write and remember
* easier to adapt existing code (it's tricky to tell which jquery utils are provided by which plugins)
* `$` means the same thing in all contexts - if it works in one place, it'll work in another.

Also, you can always fall back to the former way if you have a good reason to avoid using the monolithic `hquery`

Comment (orangejenny) [inline corehq/apps/style/static/style/js/hq_extensions.jquery.js]: There isn't a simple answer to this, and it depends on what we do with concatenating files, but it's worth remembering that a cached monolith can be better than a non-cached smaller file.

**Comment by esoergel** [comment]
I played around with this a bunch over break, but there's one fatal flaw with this approach - requirejs will make a separate request for each resource, even if it's already on the page.  What I would like to do is (ab)use requirejs to load the global resources, something like:
```javascript
var $, ko;
require(['hquery', 'knockout'], function (hquery, knockout) {
    $ = hquery;
    ko = knockout;
});
```
This works, except that the lambda function gets called asynchronously, so code on the page that may depend on `$` could get called before it's been assigned.

One approach I've been considering is to use a `synchronous_require` function I found one [here](https://github.com/scriptfoundry/require-synchronous/blob/master/require-synchronous.js), but haven't tried it out yet.  (I believe the builtin shim export config has the same sort of issue)

This is starting to feel pretty far from the spirit of require js, so I'd like to discuss options.  @millerdev Can we chat about this later?

**Comment by mwhite** [comment]
Yo yo, you should look into using [JSPM](http://jspm.io/), [SystemJS](https://github.com/systemjs/systemjs), and thereby [Babel](https://babeljs.io/). The JSPM configuration options are relatively similar to RequireJS.

**Comment by esoergel** [comment]
@mwhite Thanks for the tip, @millerdev has been looking in to alternatives to requirejs - it has a couple issues that make it less than ideal for the kind of partial, incremental migration we're attempting.  I'll read up more on JSPM/SystemJS.

**Comment by czue** [comment]
@mwhite sighting! :eyes: :wave:
